python/lsst/daf/butler/instrument.py

```diff
@@ -65,6 +71,76 @@ def _addPhysicalFilters(self, registry):
 
     def _addSensors(self, registry):
         for entry in self.sensors:
-            if "camera" not in entry:
-                entry["camera"] = self.camera
-            registry.addDataUnitEntry("Sensor", entry)
+            if 'camera' not in entry:
+                entry['camera'] = self.camera
+            registry.addDataUnitEntry('Sensor', entry)
+
+
+def makeExposureEntryFromVisitInfo(dataId, visitInfo, snap=0):
+    """Construct an Exposure DataUnit entry from `afw.image.VisitInfo`.
+
+    Parameters
+    ----------
+    dataId : `dict`
+        Dictionary of DataUnit primary/foreign key values for Exposure
+        ("camera", "exposure", optionally "visit" and "physical_filter").
+    visitInfo : `afw.image.VisitInfo`
+        A `VisitInfo` object corresponding to the Exposure.
+    snap : `int`
+        Snap index of the Exposure.
+
+    Returns
+    -------
+    entry : `dict`
+        A dictionary containing all fields in the Exposure table.
+    """
+    avg = visitInfo.getDate()
+    begin = DateTime(int(avg.nsecs(DateTime.TAI) - 0.5E9*visitInfo.getExposureTime()), DateTime.TAI)
```

Comment:
You mean we can't get the value of `DATE-OBS` from the object but need to guess it by subtracting half the exposure time?

Reply:
AFAIK, yes.  I don't know much about `VisitInfo`.  I stole this recipe from `obs.base.makeRawVisitInfo.offsetDate`, and I'd like to replace both that and `VisitInfo` itself, but this seemed like an easy way to get the correct (or at least consistent with Gen2) answer.